src/map_ref.rs

```diff
@@ -0,0 +1,214 @@
+use crate::iter::*;
+use crate::HashMap;
+use crossbeam_epoch::{self as epoch, Guard};
+use std::borrow::Borrow;
+use std::collections::hash_map::RandomState;
+use std::fmt::{self, Debug, Formatter};
+use std::hash::{BuildHasher, Hash};
+use std::ops::Index;
+
+/// A reference to a [`HashMap`], constructed with [`HashMap::as_ref`].
+/// The current thread will be pinned for the duration of this reference.
+pub struct HashMapRef<'map, K, V, S = RandomState> {
+    map: &'map HashMap<K, V, S>,
+    guard: Guard,
+}
+
+impl<K, V, S> HashMap<K, V, S> {
+    /// Get a reference to this map with the current thread pinned.
+    pub fn as_ref(&self) -> HashMapRef<'_, K, V, S> {
```

Comment:
I wonder -- do we _also_ want a constructor that _takes a `&epoch::Guard`? That'd be handy for #44